ZOOKEEPER-4210: Preserve return code from nonblocking send

smikes opened the pull request:
Async API calls attempt to flush the send buffer, which
calls flush_send_queue(); and can report
  ZOPERATIONTIMEOUT
  ZSYSTEMERROR
  ZCONNECTIONLOSS

Specifically: send_buffer() calls send(2) with MSG_NOSIGNAL,
which can return EPIPE; then send_buffer return -1, causing
ZCONNECTIONLOSS from flush_send_queue().

Current async API calls drop the return value from flush_send_queue(),
as below:

    adaptor_send_queue(zh, 0);
    return (rc < 0)?ZMARSHALLINGERROR:ZOK;

The async API then returns ZOK instead of ZCONNECTIONLOSS.

smikes commented:
Additional notes:
 - developed against 3.4 and adapted to 3.6
 - allows earlier discovery of connection loss

smikes commented:
Failure in PR-build:
```
[2021-02-12T23:45:10.488Z] [ERROR] Errors: 
[2021-02-12T23:45:10.488Z] [ERROR]   ReadOnlyModeTest.testConnectionEvents:205 » Timeout Failed to connect in read-...
[2021-02-12T23:45:10.488Z] [INFO] 
[2021-02-12T23:45:10.488Z] [ERROR] Tests run: 2901, Failures: 0, Errors: 1, Skipped: 4
```

I will try to reproduce the failure locally.

maoling commented:
@smikes 
- Thanks for this contribution. It's better to create a [JIRA](https://issues.apache.org/jira/projects/ZOOKEEPER/) issue (sign up JIRA if you don't have an account) 
to bind this PR to a JIRA-ID. ZooKeeper uses the GitHub workflow. The contributor guideline is [here](https://cwiki.apache.org/confluence/display/ZOOKEEPER/HowToContribute)
- The CI failure may be not your fault and you can ignore it:)

smikes commented:
Thanks. I created https://issues.apache.org/jira/browse/ZOOKEEPER-4210

maoling commented:
Ping @ztzg @eolivelli @anmolnar  :)

smikes commented:
Confirmed that `ReadOnlyModeTest` passes locally, so if this is known to be flaky/slow under load, then I'm fine.

smikes commented:
Running the build on branch-3.6 I see some unit tests failing, both with this change and without

ztzg reviewed (changes requested):
Hi @smikes,

Looks good, in general.  Thank you!

But…

This is technically an API & ABI break, as it is widening the range of error codes returned by API functions.  As such, I would recommend *not* including it in upcoming 3.6 releases.  (I would be okay with 3.7+ as 1/ C only uses integer codes, which most programs don't interpret and 2/ we are not on the "semantic versioning" train.  @eolivelli, WDYT?)

In any case, we usually take PRs against `master` and backport to the relevant branches; would you mind doing that?

I have also left a couple of comments on the diff, one of which is actionable.

Finally, there is the question of documenting the change in `zookeeper.h`; the comment on e.g. `zoo_aset` is now outdated and should probably be rephrased:

```c
/* [...]
 * \return ZOK on success or one of the following errcodes on failure:
 * ZBADARGUMENTS - invalid input parameters
 * ZINVALIDSTATE - zhandle state is either ZOO_SESSION_EXPIRED_STATE or ZOO_AUTH_FAILED_STATE
 * ZMARSHALLINGERROR - failed to marshall a request; possibly, out of memory
 */
ZOOAPI int zoo_aset(zhandle_t *zh, const char *path, const char *buffer, int buflen,
        int version, stat_completion_t completion, const void *data);
```

Would you agree?

Cheers, -D

P.-S. — You commented:

> Running the build on branch-3.6 I see some unit tests failing, both with this change and without

Right; we have some stability issues with the current test suite.  Passing `-Dsurefire-forkcount=1` helps with Java tests, at the cost of a looong runtime.  Something we have to fix ASAP, but please ignore CI failures which don't seem relevant in the meantime, as @maoling suggested.

smikes commented:
Sounds good, will take a look. Internally we have 3.4 and 3.6 branches active so I developed against that, I will rebase this against the main branch as you suggest

smikes commented on zookeeper-client/zookeeper-client-c/src/zookeeper.c:
```diff
@@ -3730,8 +3745,9 @@ int zoo_awget(zhandle_t *zh, const char *path,
     /* make a best (non-blocking) effort to send the requests asap */
-    adaptor_send_queue(zh, 0);
+    return nonblocking_send(zh, rc);
     return (rc < 0)?ZMARSHALLINGERROR:ZOK;
```
Fixed, see new commits.

smikes commented on zookeeper-client/zookeeper-client-c/src/zookeeper.c:
```diff
@@ -3691,6 +3691,21 @@ static int Request_path_watch_init(zhandle_t *zh, int mode,
+    } else {
+	rc = adaptor_send_queue(zh, 0);
+	if (rc < 0) {
+	    rc = handle_socket_error_msg(zh, line, rc, "Socket error during nonblocking send");
```
I added a commit that addresses this, but it is incidental to the problem of connection loss, so it may be better to put that in a different PR.  Please advise.

smikes commented:
Summary of current changes:

* bc5c997
A revised version of the original change.  This now *DOES NOT* change the ABI -- it does not modify the return values that come from the `zoo_a*` functions.

Also, the action on ZCONNECTIONLOSS is more conservative.  Now, a connection loss that is detected during an opportunistic nonblocking send will be treated as a normal connection loss, not an error.  The pending send buffers will not be cleared, and completions/watchers will be preserved at least until the next `zookeeper_interest` call.

* 9c0cd16
Addresses the comment above, that `handle_socket_error_msg` reports the caller's line but not the caller's function name.  Not needed to address ZOOKEEPER-4210, so I would be happy to split it into a different PR.  Please advise.

* e20bf70
In addition to changing the behavior when connection loss is detected, also allow the return code ZCONNECTIONLOSS to propagate upward to the `zoo_a*` functions.  Add documentation in `zookeeper.h` for the new return codes.

Note: included for completeness, but I now think this is probably a bad idea, because there is nothing the consumer of the `zookeeper.c` API can do with the new information (about connection loss during async send).

smikes commented:
Separately: I am seeing the `zktest_mt` and `zktest_st` cppunit test suites fail, it seems spuriously, on some log management calls:

```     [exec] Zookeeper_simpleSystem::testLogCallbackClearLog Message Received: [2021-02-17 17:42:23,777:207110(0x7f13ca702740):ZOO_INFO@log_env@1250: Client environment:zookeeper.version=zookeeper C client 3.8.0]
     [exec] Log Message Received: [2021-02-17 17:42:24,611:207110(0x7f13c9e8f700):ZOO_ERROR@check_events@2983: Socket 127.0.0.1:22181 zk retcode=-4, errno=111(Connection refused): server refused to accept the client]
     [exec] Log Message Received: [2021-02-17 17:42:24,611:207110(0x7f13c9e8f700):ZOO_WARN@zookeeper_interest@2561: Delaying connection after exhaustively trying all servers [127.0.0.1:22181]]
     [exec]  : elapsed 1001 : OK
     [exec] terminate called after throwing an instance of 'CppUnit::Exception'
     [exec]   what():  equality assertion failed
     [exec] - Expected: -101

ztzg commented:
@smikes: Still planning to have a look at this, and https://github.com/apache/zookeeper/pull/1609, ASAP.  Thank you for all that work so far!

ztzg reviewed (approved):
LGTM!

@eolivelli, @anmolnar, @symat: PTAL.

ztzg commented:
> Summary of current changes:
> 
> * [bc5c997](https://github.com/apache/zookeeper/commit/bc5c9975c08aa98748319ff1cd5c60b5602cba7d)
>   A revised version of the original change.  This now _DOES NOT_ change the ABI -- it does not modify the return values that come from the `zoo_a*` functions.
>   Also, the action on ZCONNECTIONLOSS is more conservative.  Now, a connection loss that is detected during an opportunistic nonblocking send will be treated as a normal connection loss, not an error.  The pending send buffers will not be cleared, and completions/watchers will be preserved at least until the next `zookeeper_interest` call.

This is this PR.  LGTM.

> * [9c0cd16](https://github.com/apache/zookeeper/commit/9c0cd1652a41269ce54120bb6656fcdd219063f7)
>   Addresses the comment above, that `handle_socket_error_msg` reports the caller's line but not the caller's function name.  Not needed to address ZOOKEEPER-4210, so I would be happy to split it into a different PR.  Please advise.

Now https://github.com/apache/zookeeper/pull/1609 and [ZOOKEEPER-4217](https://issues.apache.org/jira/browse/ZOOKEEPER-4217).  Thank you for splitting it.

I note that the PR currently is in "draft" state.  Do you have specific improvements in mind?

> * [e20bf70](https://github.com/apache/zookeeper/commit/e20bf7015ed6f28d41d488e478ec308f627fb8a7)
>   In addition to changing the behavior when connection loss is detected, also allow the return code ZCONNECTIONLOSS to propagate upward to the `zoo_a*` functions.  Add documentation in `zookeeper.h` for the new return codes.
>   Note: included for completeness, but I now think this is probably a bad idea, because there is nothing the consumer of the `zookeeper.c` API can do with the new information (about connection loss during async send).

Agree.  (But thank you for preparing the commit anyway!)

smikes commented:
> > Summary of current changes:
> > 
> > * [bc5c997](https://github.com/apache/zookeeper/commit/bc5c9975c08aa98748319ff1cd5c60b5602cba7d)
> >   A revised version of the original change.  This now _DOES NOT_ change the ABI -- it does not modify the return values that come from the `zoo_a*` functions.
> 
> This is this PR. LGTM.

Cool! I am happy to take more feedback on this, or it can be merged.

> > * [9c0cd16](https://github.com/apache/zookeeper/commit/9c0cd1652a41269ce54120bb6656fcdd219063f7)
> >   Addresses the comment above, that `handle_socket_error_msg` reports the caller's line but not the caller's function name.  Not needed to address ZOOKEEPER-4210, so I would be happy to split it into a different PR.  Please advise.
> 
> Now #1609 and [ZOOKEEPER-4217](https://issues.apache.org/jira/browse/ZOOKEEPER-4217). Thank you for splitting it.
> 
> I note that the PR currently is in "draft" state. Do you have specific improvements in mind?

No. Today I removed the "draft" status.

After this, I will make some documentation updates to the C client dev notes.

eolivelli commented on zookeeper-client/zookeeper-client-c/src/zookeeper.c:
```diff
@@ -3961,6 +3961,23 @@ static int Request_path_watch_init(zhandle_t *zh, int mode,
  * ASYNC API
  *---------------------------------------------------------------------------*/
+#define nonblocking_send(zh, rc) _nonblocking_send(zh, rc, __LINE__)
+static int _nonblocking_send(zhandle_t *zh, int rc, int line)
```
this parameter `line ` is not used, why should we declare it ?

smikes replied:
Thanks for pointing this out.  Formerly this was used because the function called handle_socket_error_msg() on error.  Since this is no longer called, we can omit the macro and the argument.  Updated PR , force-pushed amended version f0539f292

eolivelli reviewed (changes requested):
Overall looks good to me, 
but I left one question

eolivelli commented on zookeeper-client/zookeeper-client-c/src/zookeeper.c:
```diff
@@ -3961,6 +3961,22 @@ static int Request_path_watch_init(zhandle_t *zh, int mode,
+    if (rc < 0) {
+        rc = ZMARSHALLINGERROR;
+    } else {
+        rc = adaptor_send_queue(zh, 0);
```
(sorry I have one last question, I could have noticed it before).

in the previous version of this code we were basically always calling `adaptor_send_queue`, now in case of previous error (rc < 0) we are not calling that function anymore.

can this have a behavioural impact ?

smikes replied:
Thanks, this is a good question.  I have reorganized the code to minimize the change; now adaptor_send_queue() will always be called.

If an error is returned from adaptor_send_queue(), check if we are in a connected state (`is_connected`) and only then, close the socket and set state to `ZOO_NOTCONNECTED_STATE`.

Also: removed the replicated comment and collapsed the ZMARSHALLINGERROR check in `aremove_watches`.

ztzg commented on zookeeper-client/zookeeper-client-c/src/zookeeper.c:
```diff
@@ -3961,6 +3961,19 @@ static int Request_path_watch_init(zhandle_t *zh, int mode,
+/* make an attempt to send queued requests immediately without blocking */
+static int nonblocking_send(zhandle_t *zh, int rc)
+{
+    if (adaptor_send_queue(zh, 0) < 0) {
+        if (is_connected(zh)) {
+            close_zsock(zh->fd);
+            zh->state = ZOO_NOTCONNECTED_STATE;
+        }
+    }
+    return (rc < 0) ? ZMARSHALLINGERROR : ZOK;
```
@eolivelli wrote:

> in the previous version of this code we were basically always calling `adaptor_send_queue`, now in case of previous error (rc < 0) we are not calling that function anymore.
> 
> can this have a behavioural impact ?

Good point, indeed.  I don't think we have a way to be sure, so I prefer this version.

Now… sorry to be annoying, but:

I'm not sure I like the `is_connected(zh)` condition.  You previously had `zh->fd->sock != -1` here, which seems to make more sense—as the client might *not* be connected when this function is called.  (That can happen, e.g., if the client is in the middle of a `prime_connection` or a SASL authentication sequence; applications do not always wait for a state notification before they start issuing requests.)

WDYT?

smikes replied:
Sounds reasonable.  Will return to using `zh->fd != -1` as the guard for calling `close_zsock()` ; new version bc8adb0b4 is pushed.

ztzg reviewed (approved):
LGTM; thanks!

ztzg commented:
This is now in `master`.  Thank you for the contribution, and for your patience, @smikes!
